(DOCSP-16245): [iOS] SwiftUI AsyncOpen and AutoOpen docs (#1278)
mongodb/docs-realm: source/sdk/ios/examples/sync-changes-between-devices.txt

         .. literalinclude:: /examples/generated/code/start/Sync.codeblock.open-synced-realm.m
            :language: objectivec

.. _ios-open-synced-realm-swiftui:

Open a Synced Realm with SwiftUI
--------------------------------

.. versionadded:: 10.12.0

SwiftUI apps can use RealmSwift property wrappers to open {+realms+} and 

> populate views. These property wrappers provide a syntax that integrates nicely 

Reviewer: suggestion: remove editorial flavor
```suggestion
populate views. These property wrappers provide a syntax that integrates
```

with SwiftUI to let developers declare whether their client app should:

- Only open a {+realm+} if they can download updates from the server first. 
  This requires a successful network connection.
- Attempt to download updates from the server, but open a {+realm+} offline if 
  there is no network connection. This supports an offline-first approach at

>   the cost of potentially working with stale data.

Reviewer: Note: This reads more like a flowchart at the moment than describing the provided wrappers. As is, the flow I interpret is
```
    if exists(networkConnection) {
      updateFromRemote()
    } else {
      useLocalRealm()
    }
```

Author: Fair point. I omitted the bullets entirely and rewrote this intro section to be clearer, and more consistent with language used elsewhere on the page.

> {+app+}s that need to always have up-to-date information from the server, such

Reviewer: suggestion: simplify

```suggestion
{+app+}s that require up-to-date information from the server, such
```

as game apps with live leaderboards that the user can play on multiple devices, 

> should use ``@AsyncOpen`` to open a {+realm+}. This only works if the user has 

Reviewer: question: would linking AsyncOpen to the header below be useful here for users that identify this as their desired use-case?

a network connection when opening the {+realm+}, but ensures the user is never 
using the app with stale data.

{+app+}s that want to enable offline-first capability, where it's not a problem

> for the user to work with potentially stale data, should use ``@AutoOpen`` to

Reviewer: note: the phrasing for this is different than the "online-first" strategy presented above. Maybe reword to not use jargon initially (offline-first), an then use jargon.

> Apps that are tolerant of potentially stale data ...

Reviewer: same question as the other method, should there be a link here for users to jump directly to their desired use-case?

         .. literalinclude:: /examples/generated/code/start/Sync.codeblock.open-synced-realm.m
            :language: objectivec

.. _ios-open-synced-realm-swiftui:

Open a Synced Realm with SwiftUI
--------------------------------

.. versionadded:: 10.12.0

SwiftUI apps can use RealmSwift property wrappers to open {+realms+} and 
populate views. These property wrappers provide a syntax that integrates 
with SwiftUI to let developers declare whether their clients should download 
changesets before opening {+realm+}s, or open {+realm+}s with data on the 
devices.

{+app+}s that require up-to-date information from the server, such as game apps 
with live leaderboards that the user can play on multiple devices, should 
:ref:`use @AsyncOpen to open a {+realm+}<swiftui-async-open-synced-realm>`. 
This only works if the user has a network connection when opening the {+realm+}, 
but ensures the user is never using the app with stale data.

{+app+}s where it's not a problem for the user to work with potentially stale 
data, and the developer instead wants to enable offline-first capability,

> should :ref:`use @AutoOpen to open the {+realm+}<swiftui-auto-open-synced-realm>`. 

Reviewer: suggestion: This still feels a bit odd/unwieldy. I think it's because there are two clauses here but the previous use-case only has one. Maybe further reword this to introduce it like you did the previous use-case?

Author: Hmm, fair point. I've updated it again based on this feedback. Thank you!

         .. literalinclude:: /examples/generated/code/start/Sync.codeblock.open-synced-realm.m
            :language: objectivec

.. _ios-open-synced-realm-swiftui:

Open a Synced Realm with SwiftUI
--------------------------------

.. versionadded:: 10.12.0

SwiftUI apps can use RealmSwift property wrappers to open {+realms+} and 
populate views. These property wrappers provide a syntax that integrates 
with SwiftUI to let developers declare whether their clients should download 

> changesets before opening {+realm+}s, or open {+realm+}s with data on the 

Reviewer 2: (No comma when not changing the subject)
```suggestion
changesets before opening {+realm+}s or open {+realm+}s with data on the 
```

         .. literalinclude:: /examples/generated/code/start/Sync.codeblock.open-synced-realm.m
            :language: objectivec

.. _ios-open-synced-realm-swiftui:

Open a Synced Realm with SwiftUI
--------------------------------

.. versionadded:: 10.12.0

SwiftUI apps can use RealmSwift property wrappers to open {+realms+} and 
populate views. These property wrappers provide a syntax that integrates 
with SwiftUI to let developers declare whether their clients should download 
changesets before opening {+realm+}s, or open {+realm+}s with data on the 
devices.

{+app+}s that require up-to-date information from the server, such as game apps 
with live leaderboards that the user can play on multiple devices, should 
:ref:`use @AsyncOpen to open a {+realm+}<swiftui-async-open-synced-realm>`. 
This only works if the user has a network connection when opening the {+realm+}, 
but ensures the user is never using the app with stale data.

> {+app+}s where it's not a problem for the user to work with potentially stale 

Reviewer 2: I don't think this is the correct use of {+app+}? Aren't you referring to apps in general?
```suggestion
Apps where it's not a problem for the user to work with potentially stale 
```

Author: I used {+app+} here since we were talking about synced realms, but after thinking about this more, yeah, I think you're right and it should be general "Apps". Changing.

data, such as note-taking apps where users should be able to work with data 
on the device, should :ref:`use @AutoOpen to open the {+realm+}
<swiftui-auto-open-synced-realm>`. Like ``@AsyncOpen``, this attempts to 
download updates before opening the {+realm+}. However, if a network connection 
is not available, this method instead opens a {+realm+} with data on the 
device. This property wrapper enables developers to design offline-first 
capabilities into their apps.

.. _swiftui-async-open-synced-realm:

AsyncOpen
~~~~~~~~~

To download updates from your {+app+} before opening a {+realm+}, use the 
:swift-sdk:`@AsyncOpen <Structs/AsyncOpen.html>` property wrapper. This requires
the user to have a network connection.

.. code-block:: swift

   @AsyncOpen(appId: "app_id", partitionValue: <partition_value>) var asyncOpen

This SwiftUI property wrapper initiates ``Realm.asyncOpen()`` for the current 
user. The property wrapper publishes states, represented by the :swift-sdk:`AsyncOpenState 
enum <Enums/AsyncOpenState.html#/s:10RealmSwift14AsyncOpenStateO5erroryACs5Error_pcACmF>`,
which you can use to update the view. 

Pass Environment Values
```````````````````````

> From the LoginView, you can inject the ``.partitionvalue`` environment value 

Reviewer 2: ```suggestion
From the LoginView, you can inject the ``.partitionValue`` environment value 
```

> From the LoginView, you can inject the ``.partitionvalue`` environment value 

Reviewer 2: It doesn't necessarily have to come from the LoginView. Partition values may also not always be based on the user id (though probably). So, it seems a bit specific to refer to "the LoginView" in this doc context.

into the view where you are performing the ``@AsyncOpen`` to populate the 
``user.id``:

.. code-block:: swift

   struct LoginView: View {

      var body: some View {

>          ...view code...

Reviewer 2: It's always nice to try to keep it at least syntactically correct:
```suggestion
         // ...view code...
```

         AsyncOpenView()
            .environment(\.partitionValue, user.id)
      }
   }

From the view where you're performing the ``@AsyncOpen``, you can inject 
the opened {+realm+} into your content view as an environment value. The 
property wrapper uses this {+realm+} to populate the view.

.. code-block:: swift

   struct AsyncOpenView: View {

      var body: some View {
         ...view code...
         ListView()
            .environment(\.realm, realm)
      }

>    }

Reviewer 2: This block doesn't look right... where's the @AsyncOpen?

Reviewer 2: Scratch that. Self-addressed in the overall review comment.

            // The realm is currently being downloaded from the server.
            // Show a progress view.
            case .progress(let progress):
               ProgressView(progress)
            // Opening the Realm failed.
            // Show an error view.
            case .error(_):
               ErrorView()
            }
         }
      }

.. _swiftui-auto-open-synced-realm:

AutoOpen
~~~~~~~~

To open a synced {+realm+} regardless of whether the user has a network 
connection, use the :swift-sdk:`@AutoOpen <Structs/AutoOpen.html>` property 
wrapper. 

.. code-block:: swift

   @AutoOpen(appId: "app_id", partitionValue: <partition_value>) var autoOpen

This SwiftUI property wrapper attempts to initiate a ``Realm.asyncOpen()`` 
for the current user. If there is no internet connection, this property wrapper
instead returns an opened realm for the given ``appId`` and ``partitionValue``.

The property wrapper publishes states, represented by the :swift-sdk:`AsyncOpenState 
enum <Enums/AsyncOpenState.html#/s:10RealmSwift14AsyncOpenStateO5erroryACs5Error_pcACmF>`,
which you can use to update the view. 

Pass Environment Values
```````````````````````

> From the LoginView, you can inject the ``.partitionvalue`` environment value 

Reviewer 2: ```suggestion
From the LoginView, you can inject the ``.partitionValue`` environment value 
```
